kqueen/engines/openstack_kubespray.py

```diff
@@ -0,0 +1,434 @@
+                   "-o", "StrictHostKeyChecking=no",
+                   "-i", os.path.join(config.KUBESPRAY_FILES_PATH, "ssh_key"))
+
+
+def get_template_and_files():
+    # NOTE(sskripnick)
+    # we need to custom load and decode all files because of this bug:
+    # https://storyboard.openstack.org/#!/story/2002002
+    package_name = "kqueen.engines.resources.heat.kubespray"
+    files = ("master.yaml", "slave.yaml")
+    package = importlib.import_module(package_name)
+    path = list(package.__path__)[0]
+    files = {}
+    for root, dirs, _files in os.walk(path):
+        for f in _files:
+            fpath = os.path.join(root, f)
+            with open(fpath, "rb") as f:
+                files["file://" + fpath] = f.read().decode("utf8")
+    return os.path.join(path, "template.yaml"), files
+
+
+def generate_inventory(outputs, ssh_username):
+    """Generate inventory object for kubespray.
+
+    :param dict outputs: heat stack outputs
+
+    Outputs may look like this:
+    {
+        "masters": [
+            {"hostname": "host-1", "ip": "10.1.1.1"},
+            {"hostname": "host-2", "ip": "10.1.1.2"},
+            {"hostname": "host-3", "ip": "10.1.1.3"},
+        ],
+        "slaves": [
+            {"hostname": "host-4", "ip": "10.1.1.4"},
+            {"hostname": "host-5", "ip": "10.1.1.5"},
+        ],
+    }
+
+    Return value is json serializable object to be used as kubespray
+    inventory file.
+    """
+    ssh_common_args = " ".join(SSH_COMMON_ARGS)
+    conf = {
+        "all": {"hosts": {}},
+        "kube-master": {
+            "hosts": {},
+            "vars": {
+                "ansible_ssh_common_args": ssh_common_args
+            },
+        },
+        "kube-node": {"hosts": {}},
+        "etcd": {"hosts": {}},
+        "vault": {"hosts": {}},
+        "k8s-cluster": {"children": {"kube-node": None,
+                                     "kube-master": None}},
+    }
+    for master in outputs["masters"]:
+        conf["all"]["hosts"][master["hostname"]] = {
+            "access_ip": master["ip"],
+            "ansible_host": master["ip"],
+            "ansible_user": ssh_username,
+            "ansible_become": True,
+        }
+        conf["kube-master"]["hosts"][master["hostname"]] = None
+        conf["etcd"]["hosts"][master["hostname"]] = None
+        conf["vault"]["hosts"][master["hostname"]] = None
+    for slave in outputs["slaves"]:
+        conf["all"]["hosts"][slave["hostname"]] = {
+            "ansible_host": slave["ip"],
+            "ansible_user": ssh_username,
+            "ansible_become": True,
+        }
+        conf["kube-node"]["hosts"][slave["hostname"]] = None
+    ssh_args_fmt = ("-o ProxyCommand=\"ssh {}@{} "
+                    "-o UserKnownHostsFile=/dev/null "
+                    "-o StrictHostKeyChecking=no -W %h:%p\" " +
+                    ssh_common_args)
+    ssh_args = ssh_args_fmt.format(ssh_username, outputs["masters"][0]["ip"])
```

Comment:
outputs["masters"] may be empty? is it checked somewhere else?
 may be it's better to use get or check for empty outputs

Reply:
Output will always have all required keys including "masters" and "slaves". If not -- there is something exceptional happened, and exception will be raised and traceback will be logged.